File: ballast-core/npm/src/domain/events/player-added-to-vessel-role.ts
import { EventBase } from "../../messaging/event-base";
import { Game } from "../models/game";
import { Player } from "../models/player";
import { Vessel } from "../models/vessel";
import { VesselRole } from "../models/vessel-role";

export class PlayerAddedToVesselRoleDomainEvent extends EventBase {

    public static readonly id: string = "PlayerAddedToVesselRoleDomainEvent";

    public get id() {
        return PlayerAddedToVesselRoleDomainEvent.id;
    }

    public readonly gameId: string; 
    public readonly vessel: Vessel; 
    public readonly vesselRole: VesselRole; 
    public readonly player: Player; 

    private constructor(eventDateIsoString: string, gameId: string, vessel: Vessel, vesselRole: VesselRole, player: Player) {
        super(eventDateIsoString);
        this.gameId = gameId;
        this.vessel = vessel;
        this.vesselRole = vesselRole;
        this.player = player;
    }

    public static fromPlayerInGameVesselRole(game: Game, vessel: Vessel, vesselRole: VesselRole, player: Player) {
        return new PlayerAddedToVesselRoleDomainEvent(
            EventBase.getDateIsoString(),
            game.id,
            vessel,
            vesselRole,
            player
        );
    }

}

File: ballast-core/dotnet/src/application/events/PlayerAddedToVesselRole.cs
using System;
using Ballast.Core.Application.Models;
using Ballast.Core.Messaging;

namespace Ballast.Core.Application.Events
{
    public class PlayerAddedToVesselRoleEvent : EventBase, IApplicationEvent 
    {

        public static string GetId() => nameof(PlayerAddedToVesselRoleEvent);
        public override string Id => GetId();

        public Guid GameId { get; private set; }
        public VesselDto Vessel { get; private set; }
        public string VesselRole { get; private set; }
        public PlayerDto Player { get; private set; }

        private PlayerAddedToVesselRoleEvent() { } // For model binding / deserialization
        private PlayerAddedToVesselRoleEvent(string eventDateIsoString, Guid gameId, VesselDto vessel, string vesselRole, PlayerDto player) : base(eventDateIsoString)
        {
            GameId = gameId;
            Vessel = vessel;
            VesselRole = vesselRole;
            Player = player;
        }

        public static PlayerAddedToVesselRoleEvent FromPlayerInGameVesselRole(Guid gameId, VesselDto vessel, string vesselRole, PlayerDto player) =>
            new PlayerAddedToVesselRoleEvent(
                EventBase.GetDateIsoString(),
                gameId, 
                vessel, 
                vesselRole, 
                player
            );

    }
}